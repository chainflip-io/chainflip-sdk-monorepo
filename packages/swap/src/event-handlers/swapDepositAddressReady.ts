import { swappingSwapDepositAddressReady } from '@chainflip/processor/160/swapping/swapDepositAddressReady';
import { z } from 'zod';
import { calculateExpiryTime } from '../utils/function';
import { EventHandlerArgs } from './index';

const swapDepositAddressReadyArgs = swappingSwapDepositAddressReady;

export type SwapDepositAddressReadyArgs = z.input<typeof swapDepositAddressReadyArgs>;

export const swapDepositAddressReady = async ({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> => {
  const issuedBlock = block.height;

  const {
    depositAddress,
    destinationAddress,
    sourceAsset,
    destinationAsset,
    channelId,
    sourceChainExpiryBlock,
    channelMetadata,
    brokerCommissionRate,
    boostFee,
    channelOpeningFee,
    affiliateFees,
    refundParameters,
    dcaParameters,
  } = swapDepositAddressReadyArgs.parse(event.args);

  const chainInfo = await prisma.chainTracking.findFirst({
    where: {
      chain: depositAddress.chain,
    },
  });

  const estimatedExpiryTime = calculateExpiryTime({
    chainInfo,
    expiryBlock: sourceChainExpiryBlock,
  });

  const data = {
    srcChain: depositAddress.chain,
    srcAsset: sourceAsset,
    depositAddress: depositAddress.address,
    destAsset: destinationAsset,
    destAddress: destinationAddress.address,
    srcChainExpiryBlock: sourceChainExpiryBlock,
    brokerCommissionBps: brokerCommissionRate,
    maxBoostFeeBps: boostFee,
    issuedBlock,
    channelId,
    openingFeePaid: channelOpeningFee.toString(),
    fokMinPriceX128: refundParameters?.minPrice.toString(),
    fokRefundAddress: refundParameters?.refundAddress.address,
    fokRetryDurationBlocks: refundParameters?.retryDuration,
    numberOfChunks: dcaParameters?.numberOfChunks,
    chunkIntervalBlocks: dcaParameters?.chunkInterval,
    createdAt: new Date(block.timestamp),
  };

  await Promise.all([
    prisma.depositChannel.create({
      data: {
        srcChain: data.srcChain,
        issuedBlock: data.issuedBlock,
        depositAddress: data.depositAddress,
        channelId: data.channelId,
        isSwapping: true,
      },
    }),
    prisma.swapDepositChannel.upsert({
      where: {
        issuedBlock_srcChain_channelId: {
          channelId,
          issuedBlock,
          srcChain: depositAddress.chain,
        },
      },
      create: {
        affiliates: {
          createMany: {
            data: affiliateFees.map((affiliate) => ({
              account: affiliate.account,
              commissionBps: affiliate.bps,
            })),
          },
        },
        estimatedExpiryAt: estimatedExpiryTime,
        ccmGasBudget: channelMetadata?.gasBudget.toString(),
        ccmMessage: channelMetadata?.message,
        ...data,
      },
      update: data,
    }),
  ]);
};

export default swapDepositAddressReady;
