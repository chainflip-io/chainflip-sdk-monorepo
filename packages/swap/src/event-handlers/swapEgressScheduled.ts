import { z } from 'zod';
import { Asset, assetChains } from '@/shared/enums';
import { bigintMin } from '@/shared/functions';
import { chainflipChain, unsignedInteger } from '@/shared/parsers';
import { Environment, getEnvironment } from '@/shared/rpc';
import { readAssetValue } from '@/shared/rpc/utils';
import { CacheMap } from '@/swap/utils/dataStructures';
import { estimateIngressEgressFeeAssetAmount } from '@/swap/utils/fees';
import env from '../config/env';
import type { EventHandlerArgs } from '.';

const eventArgs = z.object({
  swapId: unsignedInteger,
  egressId: z.tuple([
    z.object({ __kind: chainflipChain }).transform(({ __kind }) => __kind),
    unsignedInteger,
  ]),
});

export type SwapEgressScheduledEvent = z.input<typeof eventArgs>;

const environmentByBlockHashCache = new CacheMap<
  string,
  Promise<Environment | null>
>(60_000);

const methodNotFoundRegExp = /Exported method .+ is not found/;
const rpcConfig = { rpcUrl: env.RPC_NODE_HTTP_URL };

const getCachedEnvironmentAtBlock = async (
  blockHash: string,
): Promise<Environment | null> => {
  const cached = environmentByBlockHashCache.get(blockHash);
  if (cached) return cached;

  const environment = getEnvironment(rpcConfig, blockHash).catch((e: Error) => {
    const cause = e.cause as { message: string } | undefined;

    if (methodNotFoundRegExp.test(cause?.message ?? '')) return null;

    environmentByBlockHashCache.delete(blockHash);
    throw e;
  });

  environmentByBlockHashCache.set(blockHash, environment);

  return environment;
};

const assetAmountByNativeAmountAndBlockHashCache = new CacheMap<
  string,
  Promise<bigint>
>(60_000);

const getCachedAssetAmountAtBlock = async (
  asset: Asset,
  nativeAmount: bigint,
  blockHash: string,
): Promise<bigint> => {
  const cacheKey = `${asset}-${nativeAmount.toString()}-${blockHash}`;
  const cached = assetAmountByNativeAmountAndBlockHashCache.get(cacheKey);
  if (cached) return cached;

  const rate = estimateIngressEgressFeeAssetAmount(
    nativeAmount,
    asset,
    blockHash,
  ).catch((e: Error) => {
    assetAmountByNativeAmountAndBlockHashCache.delete(blockHash);
    throw e;
  });

  assetAmountByNativeAmountAndBlockHashCache.set(cacheKey, rate);

  return rate;
};

const getEgressFeeAtBlock = async (
  blockHash: string,
  asset: Asset,
): Promise<bigint> => {
  const environment = await getCachedEnvironmentAtBlock(blockHash);
  if (!environment) return 0n;

  const nativeFee = readAssetValue(environment.ingressEgress.egressFees, {
    asset,
    chain: assetChains[asset],
  });

  return getCachedAssetAmountAtBlock(asset, nativeFee, blockHash);
};

/**
 * this event is emitted in order to correlate the egress id from a network
 * deposit/egress pallet to a swap id
 */
export default async function swapEgressScheduled({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const {
    swapId,
    egressId: [chain, nativeId],
  } = eventArgs.parse(event.args);

  const swap = await prisma.swap.findUniqueOrThrow({
    where: { nativeId: swapId },
  });

  // TODO: use an accurate source for determining the egress fee once the protocol provides one
  const egressFee = bigintMin(
    await getEgressFeeAtBlock(block.hash, swap.destAsset),
    BigInt(swap.swapOutputAmount?.toFixed() ?? 0),
  );

  await Promise.all([
    prisma.egress.update({
      where: { nativeId_chain: { chain, nativeId } },
      data: {
        amount: swap.swapOutputAmount?.sub(egressFee.toString()),
      },
    }),
    prisma.swap.update({
      where: { nativeId: swapId },
      data: {
        egress: { connect: { nativeId_chain: { chain, nativeId } } },
        fees: {
          create: {
            type: 'EGRESS',
            asset: swap.destAsset,
            amount: egressFee.toString(),
          },
        },
      },
    }),
  ]);
}
