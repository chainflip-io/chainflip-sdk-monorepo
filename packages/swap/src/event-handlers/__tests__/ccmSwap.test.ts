import { processEvents } from './utils';
import prisma from '../../client';
import { SwapDepositAddressReadyArgs } from '../swapDepositAddressReady';
import { SwapEgressScheduledArgs } from '../swapEgressScheduled';
import { SwapExecutedArgs } from '../swapExecuted';
import { SwapRequestedArgs } from '../swapRequested';
import { SwapScheduledArgs } from '../swapScheduled';

const depositChannelEvents = [
  {
    id: '0000000031-000050-2d23a',
    blockId: '0000000031-2d23a',
    indexInBlock: 50,
    extrinsicId: null,
    callId: null,
    name: 'LiquidityPools.NewPoolCreated',
    args: {
      baseAsset: { __kind: 'Dot' },
      quoteAsset: { __kind: 'Usdc' },
      initialPrice: '340282366920938470546924331736236032',
      feeHundredthPips: 20,
    },
  },
  {
    id: '0000000031-000054-2d23a',
    blockId: '0000000031-2d23a',
    indexInBlock: 54,
    extrinsicId: null,
    callId: null,
    name: 'LiquidityPools.NewPoolCreated',
    args: {
      baseAsset: { __kind: 'Flip' },
      quoteAsset: { __kind: 'Usdc' },
      initialPrice: '3402823669209384428752601088',
      feeHundredthPips: 20,
    },
  },
  {
    id: '0000000052-000001-25edd',
    blockId: '0000000052-25edd',
    indexInBlock: 1,
    extrinsicId: '0000000052-000001-25edd',
    callId: '0000000052-000001-25edd',
    name: 'PolkadotChainTracking.ChainStateUpdated',
    args: {
      newChainState: {
        blockHeight: 55,
        trackedData: {
          medianTip: '0',
          runtimeVersion: {
            specVersion: 1002005,
            transactionVersion: 26,
          },
        },
      },
    },
  },
  {
    id: '0000000052-000024-25edd',
    blockId: '0000000052-25edd',
    indexInBlock: 24,
    extrinsicId: '0000000052-000007-25edd',
    callId: '0000000052-000007-25edd',
    name: 'Swapping.SwapDepositAddressReady',
    args: {
      boostFee: 0,
      channelId: '2',
      sourceAsset: { __kind: 'Dot' },
      affiliateFees: [],
      depositAddress: {
        value: '0x7b42c78a80b6e72e1ec500cfb88a064959dfae5fe8b09f366444d35878364513',
        __kind: 'Dot',
      },
      channelMetadata: {
        message:
          '0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000003f7a000000000000000000000000000000000000000000000000000000000000000074761735465737400000000000000000000000000000000000000000000000000',
        gasBudget: '360801',
        cfParameters: '0x8ea88ab41897b921ef36ddd7dfd3e9',
      },
      destinationAsset: { __kind: 'Flip' },
      channelOpeningFee: '0',
      destinationAddress: { value: '0xa51c1fc2f0d1a1b8494ed1fe312d7c3a78ed91c0', __kind: 'Eth' },
      brokerCommissionRate: 100,
      sourceChainExpiryBlock: '157',
    } as SwapDepositAddressReadyArgs,
  },
  {
    id: '0000000057-000010-29eb0',
    blockId: '0000000057-29eb0',
    indexInBlock: 10,
    extrinsicId: '0000000057-000003-29eb0',
    callId: '0000000057-000003-29eb0',
    name: 'Swapping.SwapRequested',
    args: {
      origin: {
        __kind: 'DepositChannel',
        channelId: '2',
        depositAddress: {
          value: '0x7b42c78a80b6e72e1ec500cfb88a064959dfae5fe8b09f366444d35878364513',
          __kind: 'Dot',
        },
        depositBlockHeight: '57',
      },
      inputAsset: { __kind: 'Dot' },
      inputAmount: '499802700000',
      outputAsset: { __kind: 'Flip' },
      requestType: {
        __kind: 'Ccm',
        outputAddress: { value: '0xa51c1fc2f0d1a1b8494ed1fe312d7c3a78ed91c0', __kind: 'Eth' },
        ccmSwapMetadata: {
          swapAmounts: { principalSwapAmount: '499802700000', gasBudget: '360801' },
          depositMetadata: {
            sourceChain: { __kind: 'Polkadot' },
            channelMetadata: {
              message:
                '0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000003f7a000000000000000000000000000000000000000000000000000000000000000074761735465737400000000000000000000000000000000000000000000000000',
              gasBudget: '360801',
              cfParameters: '0x8ea88ab41897b921ef36ddd7dfd3e9',
            },
          },
        },
      },
      swapRequestId: '28',
    } as SwapRequestedArgs,
  },
  {
    id: '0000000057-000011-29eb0',
    blockId: '0000000057-29eb0',
    indexInBlock: 11,
    extrinsicId: '0000000057-000003-29eb0',
    callId: '0000000057-000003-29eb0',
    name: 'Swapping.SwapScheduled',
    args: {
      swapId: '28',
      swapType: { __kind: 'CcmPrincipal' },
      executeAt: 59,
      inputAmount: '499802339199',
      swapRequestId: '28',
    } as SwapScheduledArgs,
  },
  {
    id: '0000000057-000012-29eb0',
    blockId: '0000000057-29eb0',
    indexInBlock: 12,
    extrinsicId: '0000000057-000003-29eb0',
    callId: '0000000057-000003-29eb0',
    name: 'PolkadotIngressEgress.DepositFinalised',
    args: {
      asset: {
        __kind: 'Dot',
      },
      action: {
        __kind: 'CcmTransfer',
        swapRequestId: '28',
      },
      amount: '500000000000',
      channelId: '2',
      ingressFee: '197300000',
      blockHeight: 57,
      depositAddress: '0x7b42c78a80b6e72e1ec500cfb88a064959dfae5fe8b09f366444d35878364513',
      depositDetails: 2,
    },
  },
  {
    id: '0000000059-000036-fbdf4',
    blockId: '0000000059-fbdf4',
    indexInBlock: 36,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapExecuted',
    args: {
      swapId: '28',
      brokerFee: '4973783',
      inputAsset: { __kind: 'Dot' },
      networkFee: '497876',
      inputAmount: '499802339199',
      outputAsset: { __kind: 'Flip' },
      outputAmount: '48628752989585358147',
      swapRequestId: '28',
      intermediateAmount: '492404494',
    } as SwapExecutedArgs,
  },
  {
    id: '0000000059-000037-fbdf4',
    blockId: '0000000059-fbdf4',
    indexInBlock: 37,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapScheduled',
    args: {
      swapId: '82',
      swapType: { __kind: 'CcmGas' },
      executeAt: 61,
      inputAmount: '360801',
      swapRequestId: '28',
    } as SwapScheduledArgs,
  },
  {
    id: '0000000061-000023-df0ad',
    blockId: '0000000061-df0ad',
    indexInBlock: 23,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapExecuted',
    args: {
      swapId: '82',
      brokerFee: '0',
      inputAsset: { __kind: 'Dot' },
      networkFee: '0',
      inputAmount: '360801',
      outputAsset: { __kind: 'Eth' },
      outputAmount: '374529689743',
      swapRequestId: '28',
      intermediateAmount: '357',
    } as SwapExecutedArgs,
  },
  {
    id: '0000000061-000024-df0ad',
    blockId: '0000000061-df0ad',
    indexInBlock: 24,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapEgressScheduled',
    args: {
      asset: { __kind: 'Flip' },
      amount: '48628752989585358147',
      egressId: [{ __kind: 'Ethereum' }, '17'],
      egressFee: '374529689743',
      swapRequestId: '28',
    } as SwapEgressScheduledArgs,
  },
  {
    id: '0000000061-000025-df0ad',
    blockId: '0000000061-df0ad',
    indexInBlock: 25,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapRequestCompleted',
    args: {
      swapRequestId: '28',
    },
  },
  {
    id: '0000000061-000048-df0ad',
    blockId: '0000000061-df0ad',
    indexInBlock: 48,
    extrinsicId: null,
    callId: null,
    name: 'EthereumIngressEgress.CcmBroadcastRequested',
    args: {
      egressId: [{ __kind: 'Ethereum' }, '17'],
      broadcastId: 20,
    },
  },
  {
    id: '0000000099-000016-1f543',
    blockId: '0000000099-1f543',
    indexInBlock: 16,
    extrinsicId: '0000000099-000004-1f543',
    callId: '0000000099-000004-1f543',
    name: 'EthereumBroadcaster.BroadcastSuccess',
    args: {
      broadcastId: 20,
      transactionRef: '0x9220d619fd86e2a02d94af8c7a40333b31fdc7fe603b2ae527455a7ab1cd3364',
      transactionOutId: {
        s: '0x3232a494c1fae5495e79fd387be0f84bb158e28350d135e77f66c9e15a2ce1f2',
        kTimesGAddress: '0xf1a144d65ffed479a709d54c3ff9e906695a1d5c',
      },
    },
  },
];

const vaultOriginCcmEvents = [
  {
    id: '0002504899-000050-2d23a',
    blockId: '0002504899-2d23a',
    indexInBlock: 50,
    extrinsicId: null,
    callId: null,
    name: 'LiquidityPools.NewPoolCreated',
    args: {
      baseAsset: { __kind: 'Eth' },
      quoteAsset: { __kind: 'Usdc' },
      initialPrice: '340282366920938470546924331736236032',
      feeHundredthPips: 20,
    },
  },
  {
    id: '0002504899-000772-ae06e',
    blockId: '0002504899-ae06e',
    indexInBlock: 772,
    extrinsicId: '0002504899-000383-ae06e',
    callId: '0002504899-000383-ae06e',
    name: 'Swapping.SwapRequested',
    args: {
      origin: {
        __kind: 'Vault',
        txHash: '0x0ea0bc075d5d972ba6469bc651d2966b312085b8f7a4c55e340986c43e73abc7',
      },
      inputAsset: { __kind: 'Eth' },
      inputAmount: '1',
      outputAsset: { __kind: 'Usdc' },
      requestType: {
        __kind: 'Ccm',
        outputAddress: { value: '0x28ac2744fff7e772e6450a34d68d661a7b35c660', __kind: 'Eth' },
        ccmDepositMetadata: {
          sourceChain: { __kind: 'Ethereum' },
          sourceAddress: { value: '0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff', __kind: 'Eth' },
          channelMetadata: {
            message:
              '0x000000000000000000000000000000000000000000000000000000000000004000000000000000000000000067ff09c184d8e9e7b90c5187ed04cbfbdba741c8000000000000000000000000000000000000000000000000000000000000000c6461676f61746973686572650000000000000000000000000000000000000000',
            gasBudget: '0',
            cfParameters: '0x',
          },
        },
      },
      swapRequestId: '2448',
    },
  },
  {
    id: '0002504899-000773-ae06e',
    blockId: '0002504899-ae06e',
    indexInBlock: 773,
    extrinsicId: '0002504899-000383-ae06e',
    callId: '0002504899-000383-ae06e',
    name: 'Swapping.SwapScheduled',
    args: {
      swapId: '2401',
      swapType: { __kind: 'CcmPrincipal' },
      executeAt: 2504901,
      inputAmount: '1',
      swapRequestId: '2448',
    },
  },
  {
    id: '0002504901-000937-16632',
    blockId: '0002504901-16632',
    indexInBlock: 937,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapExecuted',
    args: {
      swapId: '2401',
      brokerFee: '0',
      inputAsset: { __kind: 'Eth' },
      networkFee: '0',
      inputAmount: '1',
      outputAsset: { __kind: 'Usdc' },
      outputAmount: '0',
      swapRequestId: '2448',
    },
  },
  {
    id: '0002504901-000938-16632',
    blockId: '0002504901-16632',
    indexInBlock: 938,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapEgressScheduled',
    args: {
      asset: { __kind: 'Usdc' },
      amount: '0',
      egressId: [{ __kind: 'Ethereum' }, '1243'],
      egressFee: '0',
      swapRequestId: '2448',
    },
  },
  {
    id: '0002504901-000939-16632',
    blockId: '0002504901-16632',
    indexInBlock: 939,
    extrinsicId: null,
    callId: null,
    name: 'Swapping.SwapRequestCompleted',
    args: { swapRequestId: '2448' },
  },
  {
    id: '0002504901-000941-16632',
    blockId: '0002504901-16632',
    indexInBlock: 941,
    extrinsicId: null,
    callId: null,
    name: 'EthereumIngressEgress.CcmBroadcastRequested',
    args: { egressId: [{ __kind: 'Ethereum' }, '1243'], broadcastId: 1929 },
  },
  {
    id: '0002504910-000016-1f543',
    blockId: '0002504910-1f543',
    indexInBlock: 16,
    extrinsicId: '0002504910-000004-1f543',
    callId: '0002504910-000004-1f543',
    name: 'EthereumBroadcaster.BroadcastSuccess',
    args: {
      broadcastId: 1929,
      transactionRef: '0x9220d619fd86e2a02d94af8c7a40333b31fdc7fe603b2ae527455a7ab1cd3364',
      transactionOutId: {
        s: '0x3232a494c1fae5495e79fd387be0f84bb158e28350d135e77f66c9e15a2ce1f2',
        kTimesGAddress: '0xf1a144d65ffed479a709d54c3ff9e906695a1d5c',
      },
    },
  },
];

describe('ccm swap flow', () => {
  beforeAll(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE public."Pool" CASCADE`;
    await prisma.pool.createMany({
      data: [
        {
          baseAsset: 'Eth',
          quoteAsset: 'Usdc',
          liquidityFeeHundredthPips: 1000,
        },

        {
          baseAsset: 'Dot',
          quoteAsset: 'Usdc',
          liquidityFeeHundredthPips: 1500,
        },
      ],
    });
  });

  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "Egress", "Broadcast", "Swap", "SwapDepositChannel", "SwapRequest", "Pool", "ChainTracking", private."State", private."DepositChannel" CASCADE`;
  });

  it('handles all the deposit channel events', async () => {
    await processEvents(depositChannelEvents);

    const fees = { select: { asset: true, amount: true, type: true } } as const;

    const channel = await prisma.swapDepositChannel.findFirstOrThrow({
      include: {
        swapRequests: {
          include: {
            fees,
            swaps: { include: { fees } },
            egress: {
              include: {
                broadcast: true,
              },
            },
          },
        },
      },
    });

    expect(channel).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      swapRequests: [
        {
          id: expect.any(BigInt),
          egressId: expect.any(BigInt),
          swapDepositChannelId: expect.any(BigInt),
          egress: {
            id: expect.any(BigInt),
            broadcastId: expect.any(BigInt),
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
            broadcast: {
              id: expect.any(BigInt),
              createdAt: expect.any(Date),
              updatedAt: expect.any(Date),
            },
          },
          swaps: [
            {
              id: expect.any(BigInt),
              swapRequestId: expect.any(BigInt),
              createdAt: expect.any(Date),
              updatedAt: expect.any(Date),
            },
            {
              id: expect.any(BigInt),
              swapRequestId: expect.any(BigInt),
              createdAt: expect.any(Date),
              updatedAt: expect.any(Date),
            },
          ],
        },
      ],
    });
  });

  it('handles all the vault events', async () => {
    await processEvents(vaultOriginCcmEvents);

    const fees = { select: { asset: true, amount: true, type: true } } as const;

    const swapRequest = await prisma.swapRequest.findUniqueOrThrow({
      where: { nativeId: 2448 },
      include: {
        fees,
        swaps: { include: { fees } },
        egress: {
          include: {
            broadcast: true,
          },
        },
      },
    });

    expect(swapRequest).toMatchSnapshot({
      id: expect.any(BigInt),
      egressId: expect.any(BigInt),
      egress: {
        id: expect.any(BigInt),
        broadcastId: expect.any(BigInt),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        broadcast: {
          id: expect.any(BigInt),
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      },
      swaps: [
        {
          id: expect.any(BigInt),
          swapRequestId: expect.any(BigInt),
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    });
  });
});
