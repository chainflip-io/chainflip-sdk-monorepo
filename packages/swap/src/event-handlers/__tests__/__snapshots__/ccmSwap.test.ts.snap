// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ccm swap flow handles all the deposit channel events 1`] = `
{
  "brokerCommissionBps": 100,
  "ccmGasBudget": "360801",
  "ccmMessage": "0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000003f7a000000000000000000000000000000000000000000000000000000000000000074761735465737400000000000000000000000000000000000000000000000000",
  "channelId": 2n,
  "chunkIntervalBlocks": null,
  "createdAt": Any<Date>,
  "depositAddress": "13ncjCBPL5daZDyC2BwKvRc63j6fNDaYCh4CD5mumGfLyatA",
  "destAddress": "0xa51c1fc2f0d1a1b8494ed1fe312d7c3a78ed91c0",
  "destAsset": "Flip",
  "estimatedExpiryAt": 1970-01-01T00:15:24.000Z,
  "expectedDepositAmount": null,
  "expiryBlock": null,
  "fokMinPriceX128": null,
  "fokRefundAddress": null,
  "fokRetryDurationBlocks": null,
  "id": Any<BigInt>,
  "isExpired": false,
  "issuedBlock": 52,
  "maxBoostFeeBps": 0,
  "numberOfChunks": 1,
  "openedThroughBackend": false,
  "openingFeePaid": "0",
  "srcAsset": "Dot",
  "srcChain": "Polkadot",
  "srcChainExpiryBlock": 157n,
  "swapRequests": [
    {
      "ccmDepositReceivedBlockIndex": "57-12",
      "ccmGasBudget": "360801",
      "ccmMessage": "0x0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000003f7a000000000000000000000000000000000000000000000000000000000000000074761735465737400000000000000000000000000000000000000000000000000",
      "completedAt": 1970-01-01T00:06:06.000Z,
      "completedBlockIndex": "61-25",
      "depositAmount": "500000000000",
      "depositBoostedAt": null,
      "depositBoostedBlockIndex": null,
      "depositReceivedAt": 1970-01-01T00:05:42.000Z,
      "depositReceivedBlockIndex": "57-12",
      "depositTransactionRef": "57-2",
      "destAddress": "0xa51c1fc2f0d1a1b8494ed1fe312d7c3a78ed91c0",
      "destAsset": "Flip",
      "effectiveBoostFeeBps": null,
      "egress": {
        "amount": "48628752989585358147",
        "broadcast": {
          "abortedAt": null,
          "abortedBlockIndex": null,
          "chain": "Ethereum",
          "createdAt": Any<Date>,
          "id": Any<BigInt>,
          "nativeId": 20n,
          "replacedById": null,
          "requestedAt": 1970-01-01T00:06:06.000Z,
          "requestedBlockIndex": "61-48",
          "succeededAt": 1970-01-01T00:09:54.000Z,
          "succeededBlockIndex": "99-16",
          "transactionRef": "0x9220d619fd86e2a02d94af8c7a40333b31fdc7fe603b2ae527455a7ab1cd3364",
          "type": "CCM",
          "updatedAt": Any<Date>,
        },
        "broadcastId": Any<BigInt>,
        "chain": "Ethereum",
        "createdAt": Any<Date>,
        "id": Any<BigInt>,
        "nativeId": 17n,
        "scheduledAt": 1970-01-01T00:06:06.000Z,
        "scheduledBlockIndex": "61-24",
        "updatedAt": Any<Date>,
      },
      "egressId": Any<BigInt>,
      "fees": [
        {
          "amount": "197300000",
          "asset": "Dot",
          "type": "INGRESS",
        },
        {
          "amount": "374529689743",
          "asset": "Flip",
          "type": "EGRESS",
        },
      ],
      "id": Any<BigInt>,
      "nativeId": 28n,
      "originType": "DEPOSIT_CHANNEL",
      "prewitnessedDepositId": null,
      "refundEgressId": null,
      "requestType": "CCM",
      "srcAddress": null,
      "srcAsset": "Dot",
      "swapDepositChannelId": Any<BigInt>,
      "swapInputAmount": "499802700000",
      "swapRequestedAt": 1970-01-01T00:05:42.000Z,
      "swaps": [
        {
          "createdAt": Any<Date>,
          "destAsset": "Flip",
          "fees": [
            {
              "amount": "497876",
              "asset": "Usdc",
              "type": "NETWORK",
            },
            {
              "amount": "9996046",
              "asset": "Dot",
              "type": "LIQUIDITY",
            },
            {
              "amount": "9848",
              "asset": "Usdc",
              "type": "LIQUIDITY",
            },
            {
              "amount": "4973783",
              "asset": "Dot",
              "type": "BROKER",
            },
          ],
          "id": Any<BigInt>,
          "intermediateAmount": "492404494",
          "latestSwapRescheduledAt": null,
          "latestSwapRescheduledBlockIndex": null,
          "nativeId": 28n,
          "retryCount": 0,
          "srcAsset": "Dot",
          "swapExecutedAt": 1970-01-01T00:05:54.000Z,
          "swapExecutedBlockIndex": "59-36",
          "swapInputAmount": "499802339199",
          "swapOutputAmount": "48628752989585358147",
          "swapRequestId": Any<BigInt>,
          "swapScheduledAt": 1970-01-01T00:05:42.000Z,
          "swapScheduledBlockIndex": "57-11",
          "type": "PRINCIPAL",
          "updatedAt": Any<Date>,
        },
        {
          "createdAt": Any<Date>,
          "destAsset": "Flip",
          "fees": [
            {
              "amount": "0",
              "asset": "Usdc",
              "type": "NETWORK",
            },
            {
              "amount": "7",
              "asset": "Dot",
              "type": "LIQUIDITY",
            },
            {
              "amount": "0",
              "asset": "Usdc",
              "type": "LIQUIDITY",
            },
          ],
          "id": Any<BigInt>,
          "intermediateAmount": "357",
          "latestSwapRescheduledAt": null,
          "latestSwapRescheduledBlockIndex": null,
          "nativeId": 82n,
          "retryCount": 0,
          "srcAsset": "Dot",
          "swapExecutedAt": 1970-01-01T00:06:06.000Z,
          "swapExecutedBlockIndex": "61-23",
          "swapInputAmount": "360801",
          "swapOutputAmount": "374529689743",
          "swapRequestId": Any<BigInt>,
          "swapScheduledAt": 1970-01-01T00:05:54.000Z,
          "swapScheduledBlockIndex": "59-37",
          "type": "GAS",
          "updatedAt": Any<Date>,
        },
      ],
    },
  ],
}
`;

exports[`ccm swap flow handles all the vault events 1`] = `
{
  "ccmDepositReceivedBlockIndex": "2504899-772",
  "ccmGasBudget": "0",
  "ccmMessage": "0x000000000000000000000000000000000000000000000000000000000000004000000000000000000000000067ff09c184d8e9e7b90c5187ed04cbfbdba741c8000000000000000000000000000000000000000000000000000000000000000c6461676f61746973686572650000000000000000000000000000000000000000",
  "completedAt": 1970-06-23T22:50:06.000Z,
  "completedBlockIndex": "2504901-939",
  "depositAmount": "1",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositReceivedAt": 1970-06-23T22:49:54.000Z,
  "depositReceivedBlockIndex": "2504899-772",
  "depositTransactionRef": "0x0ea0bc075d5d972ba6469bc651d2966b312085b8f7a4c55e340986c43e73abc7",
  "destAddress": "0x28ac2744fff7e772e6450a34d68d661a7b35c660",
  "destAsset": "Usdc",
  "effectiveBoostFeeBps": null,
  "egress": {
    "amount": "0",
    "broadcast": {
      "abortedAt": null,
      "abortedBlockIndex": null,
      "chain": "Ethereum",
      "createdAt": Any<Date>,
      "id": Any<BigInt>,
      "nativeId": 1929n,
      "replacedById": null,
      "requestedAt": 1970-06-23T22:50:06.000Z,
      "requestedBlockIndex": "2504901-941",
      "succeededAt": 1970-06-23T22:51:00.000Z,
      "succeededBlockIndex": "2504910-16",
      "transactionRef": "0x9220d619fd86e2a02d94af8c7a40333b31fdc7fe603b2ae527455a7ab1cd3364",
      "type": "CCM",
      "updatedAt": Any<Date>,
    },
    "broadcastId": Any<BigInt>,
    "chain": "Ethereum",
    "createdAt": Any<Date>,
    "id": Any<BigInt>,
    "nativeId": 1243n,
    "scheduledAt": 1970-06-23T22:50:06.000Z,
    "scheduledBlockIndex": "2504901-938",
    "updatedAt": Any<Date>,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "0",
      "asset": "Usdc",
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "nativeId": 2448n,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgressId": null,
  "requestType": "CCM",
  "srcAddress": "0xa56a6be23b6cf39d9448ff6e897c29c41c8fbdff",
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapInputAmount": "1",
  "swapRequestedAt": 1970-06-23T22:49:54.000Z,
  "swaps": [
    {
      "createdAt": Any<Date>,
      "destAsset": "Usdc",
      "fees": [
        {
          "amount": "0",
          "asset": "Usdc",
          "type": "NETWORK",
        },
        {
          "amount": "0",
          "asset": "Eth",
          "type": "LIQUIDITY",
        },
      ],
      "id": Any<BigInt>,
      "intermediateAmount": null,
      "latestSwapRescheduledAt": null,
      "latestSwapRescheduledBlockIndex": null,
      "nativeId": 2401n,
      "retryCount": 0,
      "srcAsset": "Eth",
      "swapExecutedAt": 1970-06-23T22:50:06.000Z,
      "swapExecutedBlockIndex": "2504901-937",
      "swapInputAmount": "1",
      "swapOutputAmount": "0",
      "swapRequestId": Any<BigInt>,
      "swapScheduledAt": 1970-06-23T22:49:54.000Z,
      "swapScheduledBlockIndex": "2504899-773",
      "type": "PRINCIPAL",
      "updatedAt": Any<Date>,
    },
  ],
}
`;
